flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/codegen/CodeGenUtils.scala

```diff
@@ -0,0 +1,248 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.codegen
+
+import org.apache.flink.api.common.typeinfo.PrimitiveArrayTypeInfo._
+import org.apache.flink.api.common.typeinfo.TypeInformation
+import org.apache.flink.table.`type`._
+import org.apache.flink.table.dataformat._
+import org.apache.flink.table.typeutils.TypeCheckUtils
+
+import java.lang.reflect.Method
+import java.lang.{Boolean => JBoolean, Byte => JByte, Character => JChar, Double => JDouble, Float => JFloat, Integer => JInt, Long => JLong, Short => JShort}
+import java.util.concurrent.atomic.AtomicInteger
+
+object CodeGenUtils {
+
+  // ------------------------------- DEFAULT TERMS ------------------------------------------
+
+  val DEFAULT_TIMEZONE_TERM = "timeZone"
+
+  // -------------------------- CANONICAL CLASS NAMES ---------------------------------------
+
+  val BINARY_ROW: String = className[BinaryRow]
+  val BINARY_STRING: String = className[BinaryString]
+  val BASE_ROW: String = className[BaseRow]
+  val GENERIC_ROW: String = className[GenericRow]
+
+  // ----------------------------------------------------------------------------------------
+
+  private val nameCounter = new AtomicInteger
+
+  def newName(name: String): String = {
+    s"$name$$${nameCounter.getAndIncrement}"
+  }
+
+  def newNames(names: String*): Seq[String] = {
+    require(names.toSet.size == names.length, "Duplicated names")
+    val newId = nameCounter.getAndIncrement
+    names.map(name => s"$name$$$newId")
+  }
+
+  /**
+    * Retrieve the canonical name of a class type.
+    */
+  def className[T](implicit m: Manifest[T]): String = m.runtimeClass.getCanonicalName
+
+  def needCopyForType(t: InternalType): Boolean = t match {
```

Comment:
Functions like this should better be a method of InternalType, otherwise there is no guarantee when someone add a new InternalType, he will check this functionality.
I will also open a follow up issue for this.

Reply:
+1